pallets/circuit-breaker/src/lib.rs

```diff
@@ -0,0 +1,470 @@
+use scale_info::TypeInfo;
+use sp_core::MaxEncodedLen;
+use sp_runtime::traits::{AtLeast32BitUnsigned, CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, Zero};
+use sp_runtime::{ArithmeticError, DispatchError, RuntimeDebug};
+
+pub mod weights;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarking;
+
+#[cfg(test)]
+mod tests;
+
+/// Max trade volume limit multiplier of liquidity that can be traded in a block
+pub const MAX_LIMIT_VALUE: u32 = 10_000;
+
+#[derive(Clone, Encode, Decode, RuntimeDebug, MaxEncodedLen, TypeInfo, Eq, PartialEq)]
+#[scale_info(skip_type_params(T))]
+pub struct TradeVolumeLimit<T: Config> {
+	pub amount_in: T::Balance,
+	pub amount_out: T::Balance,
+	pub limit: T::Balance,
+}
+
+impl<T: Config> TradeVolumeLimit<T>
+where
+	T::Balance: PartialOrd,
+{
+	pub fn update_amounts(&mut self, amount_in: T::Balance, amount_out: T::Balance) -> DispatchResult {
+		self.amount_in = self
+			.amount_in
+			.checked_add(&amount_in)
+			.ok_or(ArithmeticError::Overflow)?;
+		self.amount_out = self
+			.amount_out
+			.checked_add(&amount_out)
+			.ok_or(ArithmeticError::Overflow)?;
+		Ok(())
+	}
+
+	pub fn check_min_limit(&self) -> DispatchResult {
+		if self.amount_out > self.amount_in {
+			let diff = self
+				.amount_out
+				.checked_sub(&self.amount_in)
+				.ok_or(ArithmeticError::Underflow)?;
+			ensure!(diff <= self.limit, Error::<T>::MinTradeVolumePerBlockReached);
+		}
+		Ok(())
+	}
+
+	pub fn check_max_limit(&self) -> DispatchResult {
+		if self.amount_in > self.amount_out {
+			let diff = self
+				.amount_in
+				.checked_sub(&self.amount_out)
+				.ok_or(ArithmeticError::Underflow)?;
+			ensure!(diff <= self.limit, Error::<T>::MaxTradeVolumePerBlockReached);
+		}
+		Ok(())
+	}
+
+	pub fn check_limits(&self) -> DispatchResult {
+		self.check_min_limit()?;
+		self.check_max_limit()?;
+		Ok(())
+	}
+}
+
+#[derive(Clone, Encode, Decode, RuntimeDebug, MaxEncodedLen, TypeInfo, Eq, PartialEq)]
+#[scale_info(skip_type_params(T))]
+pub struct LiquidityLimit<T: Config> {
+	pub added_liquidity: T::Balance,
+	pub limit: T::Balance,
+}
+
+impl<T: Config> LiquidityLimit<T>
+where
+	T::Balance: PartialOrd,
+{
+	pub fn update_amount(&mut self, liquidity_in: T::Balance) -> DispatchResult {
+		self.added_liquidity = self
+			.added_liquidity
+			.checked_add(&liquidity_in)
+			.ok_or(ArithmeticError::Overflow)?;
+		Ok(())
+	}
+
+	pub fn check_limit(&self) -> DispatchResult {
+		ensure!(
+			self.added_liquidity <= self.limit,
+			Error::<T>::MaxLiquidityLimitPerBlockReached
+		);
+		Ok(())
+	}
+}
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use codec::HasCompact;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(_n: T::BlockNumber) {
```

Comment:
i would consider to clear values on initialize instead, given that this happens before DCA and Scheduler

Reply:
yes, but we need to figure out how much weight to reserve for it in `on_initialize`.

Reply:
you can do it ;)